server/src/handlers/http.rs

```diff
@@ -135,69 +138,111 @@ pub fn configure_routes(cfg: &mut web::ServiceConfig) {
-            web::resource("/stats").route(web::get().to(logstream::get_stats)),
+            web::resource("/stats").route(
+                web::get()
+                    .to(logstream::get_stats)
+                    .auth_stream(Action::GetStats),
+            ),
         )
         .service(
             web::resource("/retention")
                 // PUT "/logstream/{logstream}/retention" ==> Set retention for given logstream
-                .route(web::put().to(logstream::put_retention))
+                .route(
+                    web::put()
+                        .to(logstream::put_retention)
+                        .auth_stream(Action::PutRetention),
+                )
                 // GET "/logstream/{logstream}/retention" ==> Get retention for given logstream
-                .route(web::get().to(logstream::get_retention)),
+                .route(
+                    web::get()
+                        .to(logstream::get_retention)
+                        .auth_stream(Action::GetRetention),
+                ),
         );
 
     // User API
-    let user_api = web::scope("/user").service(
-        web::resource("/{username}")
-            // POST /user/{username} => Create a new user
-            .route(web::put().to(rbac::put_user))
-            // DELETE /user/{username} => Delete a user
-            .route(web::delete().to(rbac::delete_user))
-            .wrap_fn(|req, srv| {
-                // The credentials set in the env vars (P_USERNAME & P_PASSWORD) are treated
-                // as root credentials. Any other user is not allowed to modify or delete
-                // the root user. Deny request if username is same as username
-                // from env variable P_USERNAME.
-                let username = req.match_info().get("username").unwrap_or("");
-                let is_root = username == CONFIG.parseable.username;
-                let call = srv.call(req);
-                async move {
-                    if is_root {
-                        return Err(ErrorBadRequest("Cannot call this API for root admin user"));
-                    }
-                    call.await
+    let user_api = web::scope("/user")
+        .service(
+            web::resource("/{username}")
+                // PUT /user/{username} => Create a new user
+                .route(web::put().to(rbac::put_user).auth(Action::PutUser))
+                // DELETE /user/{username} => Delete a user
+                .route(web::delete().to(rbac::delete_user).auth(Action::DeleteUser))
+                .route(web::put().to(rbac::put_roles).auth(Action::PutUser)),
+        )
+        .service(
+            web::resource("/{username}/roles")
+                // PUT /user/{username}/roles => Put roles for user
+                .route(web::put().to(rbac::put_roles).auth(Action::PutRoles)),
+        )
+        .wrap_fn(|req, srv| {
+            // The credentials set in the env vars (P_USERNAME & P_PASSWORD) are treated
```

Comment:
Can we separate this in a separate private function - easier to read

Reply:
Can change this to wrap .. but that will require more boilerplate for this middleware. ( similar to middleware defined in auth_middleware.rs

Reply:
Still okay because it will make readability much better